### ISSUE TITLE
[v4] Supporting browser modules

### ISSUE THREAD

**Issue by thecodingdude** [twbs/bootstrap#25246]
Hello fine folks of Bootstrap, today I wanted to discuss the potential of modifying the bootstrap source JS to support browser based imports natively, without any third party tools/transpilers or the like.

## Preface

At the moment, all of [Bootstraps JS modules ](https://github.com/twbs/bootstrap/tree/v4-dev/js/src) are wirtten to support ES6 and export the class definition. This issue would require a slight tweak so that they are browser complaint, additionally, the import $ from jQuery declaration would need to be removed.

The upside of this, is that code such as 

import {Alert} from './js/bootstrap/alert.js';

would now work when using <script type='module'>, which even though has somewhat [limited](https://caniuse.com/#search=modules) scope, still becomes a useful feature in the future. Especially now Chrome has added [dynamic imports](https://developers.google.com/web/updates/2017/11/dynamic-import).

## Changes required

The changes are trivial:

1: Delete import $ from 'jquery' from each src module file

2: 
import Alert from './alert' 
becomes 
import Alert from './alert.js'

3: In your HTML file:

```
<script type="module">

//import {Alert} from './js/bootstrap/index.js';
import Alert from './js/bootstrap/alert.js';

//$(".alert").alert('close');

</script>
```

You can also use dynamic imports using this method too, although I have not personally tested this.

## Performance benefits

bootstrap.min.js (2298 lines) - 14.9kb
util.js (162 lines) - 4.2kb (unminified), **1.7kb minified**
alert.js (193 lines) - 4.5kb (unminified), **1.7kb minified**

Loading _only_ what you need, and loading _on demand_ has obvious advantages and is something that is possible today :)

## Next steps

Since the changes are fairly trivial, is it possible to integrate these into the v4 branch? I'm happy to PR some documentation changes. If it's possible to make this browser-module compatible without much effort that would be awesome. Heck, maybe they can be exported in a browser-module friendly format during build time.

Cheers.

**Comment by Johann-S** [comment]
Hi @thecodingdude ,

I agree this will be very usefull in the futur, for now we have to support IE 10 so I'm not sure we will work on that for this version, but we will think of it for our next major release for sure 👍 

Our plugins are built separately into an IIFE and they can be imported in your HTML to just load what you need : 

```html
<script src="node_modules/bootstrap/js/dist/util.js"></script>
<script src="node_modules/bootstrap/js/dist/modal.js"></script>

**Comment by thecodingdude** [comment]
Hi @Johann-S, thanks for looking into this. Although I must say I'm bitterly disappointed you've decided to just close this issue without further thought.

I understand bootstrap wants to target IE11 - this change does _not_ propose changes that would break IE11 support, but rather, for those that do not care, can decide at their own leisure to use modules rather than importing the entire file at load time.

I'm not proposing anything more than 2 simple modifications to your ES6-ready code to enable them to be used in the browser with some documentation updates - more ways to do something that already exist can't be a bad thing right?

**Comment by Johann-S** [comment]
Can you submit a PR ? And we will see 👍

**Comment by thecodingdude** [comment]
I'm not sure _what_ exactly I'm submitting a PR _for_ though. Hence why I made the issue before working on it in more depth.

If it's possible, the easiest and most ideal is to export the code in a browser-module friendly way, rather than making the developer do the changes themselves, they are fairly trivial, but will take time and becomes another chore, if it's possible to export them as browser-module friendly, that's the best approach.

Alternatively, it is possible just to add some docs explaining the changes required. 

All depends on what the community feels is the best approach to this, hence why I made the issue :)
